### PR TITLE
Training pipeline changed to support fp16 based training

### PR DESCRIPTION
**Pull request by ihavnoid**
See parse.py main() function to change float16/float32 training

### CODE REVIEW

File: training/tf/tfprocess.py [hunk 1]
```diff
@@ -21,32 +21,47 @@
...
-    initial = tf.constant(0.0, shape=shape)
+def bn_bias_variable(shape, dtype=tf.float32):
+    initial = tf.constant(0.0, shape=shape, dtype=dtype)
     return tf.Variable(initial, trainable=False)
 
 def conv2d(x, W):
     return tf.nn.conv2d(x, W, data_format='NCHW',
                         strides=[1, 1, 1, 1], padding='SAME')
 
+# copied from http://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/
```
[the remark this answers is not in the document]
Comment (roy7) [reply training/tf/tfprocess.py]: There's a very similar wrapped used here https://github.com/aymericdamien/TensorFlow-Examples/blob/master/examples/3_NeuralNetworks/convolutional_network_raw.py#L37 where strides=1 and so could be dropped, and you aren't using bias_add so the 'b' param could be dropped.

I can't find any copyright info from that guy, although it's example TF code so I assume it must be intended for people to copy and use. Hmm.

[the remark this answers is not in the document]
Comment (roy7) [reply training/tf/tfprocess.py]: Ah, the example I linked uses MIT License. https://github.com/aymericdamien/TensorFlow-Examples/blob/master/LICENSE

[the remark this answers is not in the document]
Comment (roy7) [reply training/tf/tfprocess.py]: Apologies if you aren't talking about the conv2d wrapper.

**Comment by ihavnoid** [comment]
I thought this was something nvidia intended other people to reuse things but I think I should be a bit more cautious.  Point taken.  I can't find any alternative implementation though.  I think the next thing I can do is to implement something similar myself.  They aren't going to claim copyright violation because I used their idea which they made public anyway, correct?

**Comment by ihavnoid** [comment]
Anyway I will re-submit something after rewriting the custom getter.

**Pull request closed without merging**